packages/file_selector/file_selector_windows/lib/src/file_selector_dart/dialog_wrapper.dart

```diff
@@ -0,0 +1,202 @@
+  void setFileTypeFilters(List<XTypeGroup> filters) {
+    final Map<String, String> filterSpecification = <String, String>{};
+
+    if (filters.isEmpty) {
+      filterSpecification[_allowAnyValue] = _allowAnyExtension;
+    } else {
+      for (final XTypeGroup option in filters) {
+        final String? label = option.label;
+        if (option.allowsAny || option.extensions!.isEmpty) {
+          filterSpecification[label ?? _allowAnyValue] = _allowAnyExtension;
+        } else {
+          final String extensionsForLabel = option.extensions!
+              .map((String extension) => '*.$extension')
+              .join(';');
+          filterSpecification[label ?? extensionsForLabel] = extensionsForLabel;
+        }
+      }
+    }
+
+    using((Arena arena) {
+      final Pointer<COMDLG_FILTERSPEC> registerFilterSpecification =
+          arena<COMDLG_FILTERSPEC>(filterSpecification.length);
+      int index = 0;
+      for (final String key in filterSpecification.keys) {
+        registerFilterSpecification[index]
+          ..pszName = key.toNativeUtf16(allocator: arena)
+          ..pszSpec = filterSpecification[key]!.toNativeUtf16(allocator: arena);
+        index += 1;
+      }
+
+      _lastResult = _dialogController.setFileTypes(
+        filterSpecification.length,
+        registerFilterSpecification,
+      );
+    });
+  }
+
+  /// Displays the dialog, and returns the selected files, or null on error.
+  List<String?>? show(int parentWindow) {
+    _lastResult = _dialogController.show(parentWindow);
+    if (!SUCCEEDED(_lastResult)) {
+      return null;
+    }
+    late List<String>? files;
+
+    using((Arena arena) {
+      final Pointer<Pointer<COMObject>> shellItemArrayPtr =
+          arena<Pointer<COMObject>>();
+      final Pointer<Uint32> shellItemCountPtr = arena<Uint32>();
+      final Pointer<Pointer<COMObject>> shellItemPtr =
+          arena<Pointer<COMObject>>();
+
+      files =
+          _getFilePathList(shellItemArrayPtr, shellItemCountPtr, shellItemPtr);
+    });
+    return files;
+  }
+
+  List<String>? _getFilePathList(
+      Pointer<Pointer<COMObject>> shellItemArrayPtr,
+      Pointer<Uint32> shellItemCountPtr,
+      Pointer<Pointer<COMObject>> shellItemPtr) {
+    final List<String> files = <String>[];
+    if (_isOpenDialog) {
+      _lastResult = _dialogController.getResults(shellItemArrayPtr);
+      if (!SUCCEEDED(_lastResult)) {
+        return null;
+      }
+
+      final IShellItemArray shellItemResources =
+          IShellItemArray(shellItemArrayPtr.cast());
+      _lastResult = shellItemResources.getCount(shellItemCountPtr);
+      if (!SUCCEEDED(_lastResult)) {
+        return null;
+      }
+      for (int index = 0; index < shellItemCountPtr.value; index += 1) {
+        shellItemResources.getItemAt(index, shellItemPtr);
+        final IShellItem shellItem = IShellItem(shellItemPtr.cast());
+        files.add(_shellWin32Api.getPathForShellItem(shellItem));
```

Comment:
`ShellWin32Api.getPathForShellItem` returns an empty string on failure. How should we handle that? What are your thoughts on making `ShellWin32Api.getPathForShellItem` throw a `WindowsException` on failure?